### PR TITLE
feat(ADDON-88124): add IndexName validator for Splunk index name validation

### PR DESCRIPTION
**Pull request by okashaev-splunk**
## Why

This PR is part of a multilayer change to replace current per-add-on length and regex index name validators with a central, authoritative validation endpoint. The new validation also covers MDL (Machine Data Lake) federated dataset names — `~.federated.<name>` — which route data to Machine Data Lake via Ingest Processor and have naming rules that a single regex cannot handle correctly.

The changes:
- **SplCore** — implements `POST /services/data/indexes/_validateName` as the single point of truth for index name validation. All future rule changes happen here only, without requiring updates across individual add-ons.
- **splunktaucclib (this PR)** — adds `IndexName` validator that calls the SplCore endpoint. Falls back to a local reimplementation of the same rules to stay compatible with older Splunk versions where the endpoint is not yet available.
- **splunk-add-on-ucc-framework** — adds `{"type": "index_name"}` support to `globalConfig.json` schema and generates `validator.IndexName()` in REST handler code. Most add-ons have no custom REST handler and rely entirely on ucc-gen; without this change they would need to add a custom handler just to use the new validator.
- **Technology add-ons** — replace current length and regex validators in `globalConfig.json` with `{"type": "index_name"}`. Add-ons with existing custom validators replace those with `IndexName` directly.

## What

Adds a new `IndexName` validator class to `splunktaucclib/rest_handler/endpoint/validator.py`.

## How

- Calls `POST /services/data/indexes/_validateName` (Splunk core endpoint) to validate a candidate index name without creating anything.
- Session key is auto-detected from `__main__.___sessionKey` (written by `MConfigHandler.__init__` from stdin before any handler method runs).
- splunkd URI is auto-detected from the `SPLUNKD_URI` environment variable, falling back to `solnlib.splunkenv.get_splunkd_uri()`.
- Falls back to local pure-Python validation rules (mirroring `IndexAdminHandler::validateName()` from splcore develop branch) when the endpoint is unavailable: HTTP 404 (older Splunk versions), HTTP 400 with unsupported `index_name` argument, or no session key available.
- `_call_validate_endpoint` returns `(None, None)` to signal fallback, `(True, None)` for valid, `(False, str)` for invalid or request failure.
- `validate_fallback` and helpers (`_validate_as_index_name`, `_validate_as_dataset_name`) return a plain `str` — empty means valid, non-empty is the error message.
- Handles three name shapes: regular index names, `federated:` prefix (stripped before applying index rules), and `~.federated.` MDL dataset names (routed to dataset-specific validation).
- `ucclog` logging throughout `_call_validate_endpoint` (WARNING for 404/400/403/5xx, ERROR for unexpected exceptions and malformed responses) and `validate()` (DEBUG for local fallback rejections).
- `self._msg` cleared at the start of `validate()` to prevent stale error messages when the validator instance is reused across requests.
- Lazy import of `ucclog` inside methods to avoid `solnlib.log.Logs()` being called at import time (requires Splunk runtime).
- Endpoint response parsing and HTTP error message extraction both wrapped defensively: `except (KeyError, IndexError, TypeError, ValueError)` covers missing keys, empty/null lists, null content, and non-JSON bodies.
- Zero-argument constructor — works as a drop-in `RestField` validator at class definition time, consistent with all existing validators.
- `_INTERNAL_INDEXES` frozenset sourced from splcore develop branch `IndexAdminHandler::validateName()`.

## Usage

```python
RestField("index", required=True, validator=IndexName())
```

## Tests

98 unit tests across five test classes:

- `TestValidateFallback` — parametrized valid/invalid name cases for local rules, covering regular names, `federated:` prefix, `~.federated.` MDL dataset names, case-insensitive `kvstore` and `default` checks, length boundaries, and all reserved internal index names.
- `TestGetSessionKey` — reads `__main__.___sessionKey`; absent attribute returns `None`.
- `TestGetSplunkdInfo` — `SPLUNKD_URI` env var parsing; solnlib fallback when env var absent.
- `TestCallValidateEndpoint` — HTTP interaction: correct `client.post()` endpoint/body, correct `SplunkRestClient` constructor args, 404/403/400/5xx handling with log assertions, malformed response shapes (non-JSON, empty entry list, missing content key, `content: null`, `messages: null`, empty `messages` list), no session key path.
- `TestValidate` — orchestration: endpoint called with correct value, fallback triggered on `(None, None)`, `self.msg` set correctly on failure, no-session-key falls back to local rules.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 1]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+
+        from solnlib.splunkenv import get_splunkd_uri
+
+        splunkd_uri = os.environ.get("SPLUNKD_URI") or get_splunkd_uri()
+        return urllib.parse.urlparse(splunkd_uri)
+
+    @staticmethod
+    def _extract_http_error_message(exc):
+        try:
+            return json.loads(exc.body).get("messages", [{}])[0].get("text")
```
Comment (wtobis-splunk) [inline splunktaucclib/rest_handler/endpoint/validator.py]: This still lets some malformed error responses escape as exceptions. For example, if splunkd returns `{"messages": []}`, `[0]` raises `IndexError`, so `validate()` can throw while handling `binding.HTTPError` instead of returning a controlled validation error. Please parse this defensively, or at least include `IndexError`/`TypeError`/`AttributeError` in the fallback path and add a regression test for an empty `messages` list.

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 2]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+                False,
+                f"Index name validation request failed, unexpected error: {exc}",
+            )
+
+        try:
+            content = json.loads(response.body.read())["entry"][0]["content"]
+        except (KeyError, IndexError, ValueError) as exc:
+            return False, f"Unexpected response from validation endpoint: {exc}"
+
+        if content["is_valid"] == "true":
```
Comment (wtobis-splunk) [inline splunktaucclib/rest_handler/endpoint/validator.py]: This access sits outside the malformed-response guard above. A response with `content` present but missing `is_valid` (for example `{"entry": [{"content": {"name": "main"}}]}`) raises `KeyError` here, so the validator leaks an exception even though other malformed shapes return `Unexpected response from validation endpoint`. Please read `is_valid` defensively, or keep this check inside the `try` block, and add coverage for a missing `is_valid` key.

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 3]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+                "non-empty and cannot start with '_' or '-'",
+            )
+
+        if not IndexName._VALID_CHARS.match(src):
+            return (
+                False,
+                "Invalid name: only lower case alphanumeric characters, '-' and '_' are allowed.",
+            )
+
+        if src_lower == "kvstore":
```
Comment (artemrys) [inline splunktaucclib/rest_handler/endpoint/validator.py:571]: Why special case for `kvstore`?

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 4]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+
+        if len(src) > IndexName._INDEX_NAME_MAX_LEN:
+            return False, f"index name={src} is too long, please try a shorter name"
+
+        if not src or src[0] == "-":
+            return (
+                False,
+                f"invalid name: '{src}'. name parameter must be "
+                "non-empty and cannot start with '_' or '-'",
+            )
```
Comment (artemrys) [inline splunktaucclib/rest_handler/endpoint/validator.py]: the condition and error message are not in sync

if `src=="_internal"` - the error message will say "blabla cannot start with '_' or '-'"

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 5]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+    >>> from splunktaucclib.rest_handler.endpoint.validator import IndexName
+    >>>
+    >>> RestField("index", required=True, validator=IndexName())
+
+    """
+
+    _ENDPOINT = "data/indexes/_validateName"
+    _INDEX_NAME_MAX_LEN = 2048
+    _DATASET_NAME_MAX_LEN = 1024
+    _INTERNAL_INDEXES = frozenset(
```
Comment (artemrys) [inline splunktaucclib/rest_handler/endpoint/validator.py:515]: I assume those internal indexes are different per Splunk version? Which version does this represent?

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 6]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+    _MSG_INDEX_NAME_NOT_SUPPORTED = (
+        'Argument "index_name" is not supported by this handler.'
+    )
+    # SplunkCore index services prefix — stripped before applying IndexName rules
+    _FEDERATED_INDEX_PREFIX = "federated:"
+    # MDL prefix — routes to dataset name validation instead
+    _FEDERATED_MDL_PREFIX = "~.federated."
+
+    @staticmethod
+    def _validate_as_index_name(name: str) -> tuple:
```
Comment (artemrys) [inline splunktaucclib/rest_handler/endpoint/validator.py]: Do we need to return tuple?

it always returns `False and text` OR `True and empty text`, can it be simplified to just return `text` or `None`?

Comment (artemrys) [reply splunktaucclib/rest_handler/endpoint/validator.py]: same for `_validate_as_dataset_name`

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 7]
```diff
@@ -483,3 +484,232 @@ def validate(self, value, data):
...
+        try:
+            content = json.loads(response.body.read())["entry"][0]["content"]
+        except (KeyError, IndexError, ValueError) as exc:
+            return False, f"Unexpected response from validation endpoint: {exc}"
+
+        if content["is_valid"] == "true":
+            return True, None
+        return False, content.get("reason", "Invalid index name")
+
+    def validate(self, value, data):
```
Comment (artemrys) [inline splunktaucclib/rest_handler/endpoint/validator.py:713]: We are missing logging here completely, if incident occurs - how would you know what logic was hit?

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 5 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:515]: They are taken from current index validation implementation in splcore

Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:515]: ... splcore develop branch

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 3 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:571]: it's reserved name and it does not start with underscore so requires separate check

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 1 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Fixed: added `IndexError` (for empty `messages` list) and `AttributeError` (for non-dict JSON body) to the `except` tuple in `_extract_http_error_message`. Also added a regression test for the empty `messages` list case (`test_400_empty_messages_list_returns_false_with_status`).

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 2 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Confirmed fixed: `content["is_valid"]` is inside the `try` block, so a missing `is_valid` key is caught by the existing `except (KeyError, IndexError, ValueError)` handler and returns `(None, None)` triggering local fallback.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 4 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Fixed: the error message now reads `"Index name cannot start with '_' unless it's one of reserved internal indexes: ..."`, which is accurate — it only fires for names starting with `_` that are not in the allowed set.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 5 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:515]: Added a comment to the constant: `# Internal indexes from splcore develop branch IndexAdminHandler::validateName()`.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 6 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Fixed: `_validate_as_index_name`, `_validate_as_dataset_name`, and `validate_fallback` all now return plain `str` — empty string means valid, non-empty string is the error message.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 7 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:713]: Fixed: `validate()` now logs at `DEBUG` level when local fallback rejects a name: `ucclog.logger.debug("IndexName: local validation rejected '%s': %s", value, error)`.

Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:713]: added 8 log messages in total, most of them in _call_validate_endpoint to reflect endpoint response

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 3 above)
Comment (artemrys) [reply splunktaucclib/rest_handler/endpoint/validator.py:571]: there are index names in `_INTERNAL_INDEXES` that do not start with underscore as well

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 7 above)
Comment (artemrys) [reply splunktaucclib/rest_handler/endpoint/validator.py:713]: is this DEBUG log visible when you switch to DEBUG logs in the TA configuration itself?

in other words, what needs to be done to see those logs?

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 3 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:571]: sound like an issue in splcore index name validation, indexes without underscore are ignore by if statement.

but the main reason kvstore is not in the list - _INTERNAL_INDEXES used as white list, while kvstore is forbidden name

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 8]
```diff
@@ -483,3 +484,248 @@ def validate(self, value, data):
...
+        from solnlib.splunkenv import get_splunkd_uri
+
+        splunkd_uri = os.environ.get("SPLUNKD_URI") or get_splunkd_uri()
+        return urllib.parse.urlparse(splunkd_uri)
+
+    @staticmethod
+    def _extract_http_error_message(exc):
+        try:
+            return json.loads(exc.body).get("messages", [{}])[0].get("text")
+        except (AttributeError, IndexError, KeyError, ValueError):
```
Comment (wtobis-splunk) [inline splunktaucclib/rest_handler/endpoint/validator.py]: Thanks, this covers the empty `messages` list and non-dict response bodies, but one malformed shape still escapes: `{"messages": null}` raises `TypeError` at the `[0]` access, and `TypeError` is not in this tuple. Since the earlier thread also called out `TypeError`, please include it here and add a small regression case for `messages: null`.

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 9]
```diff
@@ -483,3 +484,248 @@ def validate(self, value, data):
...
+                f"Index name validation request failed, unexpected error: {exc}",
+            )
+
+        try:
+            response_body = response.body.read()
+            content = json.loads(response_body)["entry"][0]["content"]
+            if content["is_valid"] == "true":
+                return True, None
+            return False, content.get("reason", "Invalid index name")
+        except (KeyError, IndexError, ValueError) as exc:
```
Comment (wtobis-splunk) [inline splunktaucclib/rest_handler/endpoint/validator.py]: Moving `content["is_valid"]` inside the `try` fixes the missing-key case you mentioned, but a type-invalid content shape can still leak. For example, `{"entry": [{"content": null}]}` raises `TypeError` at `content["is_valid"]`, which this handler does not catch. Please either catch `TypeError` here or validate that `content` is a dict before reading it, and add coverage for that malformed response shape.

File: splunktaucclib/rest_handler/endpoint/validator.py [hunk 10]
```diff
@@ -483,3 +484,248 @@ def validate(self, value, data):
...
+            )
+            return (
+                False,
+                f"Index name validation request failed, unexpected error: {exc}",
+            )
+
+        try:
+            response_body = response.body.read()
+            content = json.loads(response_body)["entry"][0]["content"]
+            if content["is_valid"] == "true":
```
Comment (wtobis-splunk) [inline splunktaucclib/rest_handler/endpoint/validator.py:704]: Can you confirm the real `_validateName` response returns string values for `is_valid` rather than JSON booleans? The tests mock `"true"` / `"false"`, but if splunkd returns boolean `true`, this comparison rejects a valid name. If the endpoint contract is definitely string-based, it would be useful to capture that with a fixture/comment from the real response shape.

Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:704]: image

Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:704]: Confirmed: the endpoint returns string `"true"` / `"false"`, not JSON booleans. Splunk REST API uses `output_mode=json` but field values in the response body are still serialized as strings for this endpoint. Using `bool()` on the value would be unsafe — any non-empty string (including `"false"`) is truthy in Python. Broadening to `in ("true", "1", True, 1)` adds complexity for cases that cannot occur under the documented contract. `== "true"` is kept as-is.

Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:704]: https://splunk.atlassian.net/wiki/spaces/PROD/pages/1080115200419/Installation+and+Testing+of+Splunk+Feature+Build+with+Index+Name+Validation

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 8 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Fixed: added `TypeError` to `_extract_http_error_message`. Also simplified to direct key access `["messages"][0]["text"]` removing the `.get()` wrappers — `AttributeError` dropped, `TypeError` covers both the `null` value and array-shaped body. Regression test added for `messages: null`.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 9 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py]: Fixed: added `TypeError` to `except (KeyError, IndexError, TypeError, ValueError)` in the response parsing block. Regression test added for `{"entry": [{"content": null}]}`.

File: splunktaucclib/rest_handler/endpoint/validator.py (hunk 7 above)
Comment (okashaev-splunk) [reply splunktaucclib/rest_handler/endpoint/validator.py:713]: Yes — `ucclog.logger` is backed by `solnlib.log.Logs().get_logger("ucc_lib")`. When you switch to DEBUG in the TA configuration UI, the TA calls `set_log_level("DEBUG")` → `solnlib.log.Logs().set_level(logging.DEBUG)`, which propagates to the `ucc_lib` logger. So all `DEBUG` messages from the validator will appear in the TA's log file (`ta_<name>.log`) at that point. No extra configuration is needed.

**Review by artemrys — APPROVED** [review]
